docs/jsx.md

```diff
@@ -134,4 +136,81 @@ You're effectively passing the array `[| <div />, <div /> |]` to `MyReasonCompon
 <MyReasonComponent> <div /> </MyReasonComponent>
 ```
 
-We unwrap this for you automatically to just `<div />` instead of an array of a single element.
\ No newline at end of file
+We unwrap this for you automatically to just `<div />` instead of an array of a single element.
+
+## Migrating from version 2 to version 3
+
+There are many ReasonReact applications, so it is hard to define "The One True" migration strategy for them all.
+
+Depending on the size and nature of your application there are two options available to migrate from JSX version 2 to version 3.
+
+### Application level
+
+By adding `{"reason": {"react-jsx": 3}` in your [`bsconfig.json`](https://bucklescript.github.io/docs/en/build-configuration.html#reason-refmt).
+
+This approach requires that all components in the application must be made compatible with version 3 of JSX at once, so it will be a better fit for smaller apps with a reduced number of components, where all of them can be migrated to version 3 in one fell swoop.
+
+### File level
+
+For larger applications, it might not be possible to migrate all components at once. In these cases, a per-file migration is also possible.
+
+A file can be configured to use version 3 of the transform by adding `[@bs.config {jsx: 3}]` at the top of the file.
+
+The per-file configuration allows to mix, in the same application, components compatible with either of both versions of the JSX transforms. However, the restriction is that all the components used in a file will have to be compatible with the JSX version specified for that file.
+
+For example, if a file contains the following code:
+
+```reason
+/* User.re */
+[@bs.config {jsx: 3}]
+[@react.component]
+let make = (~id) => {
+  <Profile>
+    <UserDetails id />
+  </Profile>;
+};
+```
+
+Then `Profile` and `UserDetails` components will have to be compatible with the version 3 of JSX. Or alternatively, if they are using version 2, they can be wrapped with the function `ReasonReactCompat.wrapReasonReactForReact`.
+
+#### From primitives to more complex components
+
+As all usages of any component in a file need to be migrated to the same version of JSX, one natural way to tackle large migrations at the file level is to start converting the most primitive components to version 3, as they generally render elements of a reduced number of components, or host elements like `<div />`. Once the most simple components are done, one can proceed with the most complex ones.
+
+While this process is ongoing, both versions of JSX need to coexist. For example, a component `Banner` compatible with version 2 might need to use a component `Image` that is compatible with version 3. `Image` can be made compatible by leveraging `ReasonReactCompat.wrapReactForReasonReact`:
+
+```reason
+/* In Image.re */
+[@bs.config {jsx: 3}];
+
+[@react.component]
+let make = (~src) => <img src />;
+
+module Jsx2 = {
+  let component = ReasonReact.statelessComponent("Image");
+  let make = (~src, children) =>
```

Comment:
@rickyvetter I'm not sure if `key` has to be added here?

Reply:
Nope. Not needed!

Reply:
@jchavarri @rickyvetter rightfully insisted (Discord) that we added ~ to children.

Reply:
@diligiant In this case—nope, since it's a wrapper to JSX@2 API: it requires unlabeled last argument.

Reply:
@alexfedoseev ok ; don't you think we need a section about these precise "rules" ? Once you know, you know but we might end up with troubles for newbies unless the rls emits a warning/error.

Reply:
@diligiant I added an inline comment in https://github.com/reasonml/reason-react/pull/391/commits/36a4c98858c2077d9421bb22cea3913ae40efbe8 that explains why the label is not needed here. Please let me know if you have any specific suggestions on how to make it better.

Reply:
@jchavarri Thank you. It goes back to the difference with React where children is a special prop (3rd param of createElement) ; @rickyvetter was kind enough to highlight this to me.

Likewise, I don't think I've seen the explanation that `=?` inside an interop external is like v2 `[@bs.optional]`, different from normal `=?` (turns the param into `option(a')`).